### PR TITLE
Disallow functions in react-redux#connect()

### PR DESCRIPTION
**Pull request by mcsf**
See https://github.com/Automattic/wp-calypso/issues/14024

My first eslint rule, so be gentle. 😄

### CODE REVIEW

File: lib/rules/redux-no-bound-selectors.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
+/**
+ * @fileoverview Disallow the use of this.translate
```
Comment (mcsf) [inline lib/rules/redux-no-bound-selectors.js]: Oopsies, forgot this

File: lib/rules/redux-no-bound-selectors.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+// Rule Definition
+//------------------------------------------------------------------------------
+
+const REF = ' See wp-calypso#14024';
+const BIND_ERROR_MESSAGE = "Don't bind functions within `connect`." + REF;
+const FUNC_ERROR_MESSAGE = "Don't instantiate functions within `connect`." + REF;
+
+function isStateInArgs( node ) {
+	return node.params.some( function( paramNode ) {
+		return paramNode.name === 'state';
```
Comment (aduth) [inline lib/rules/redux-no-bound-selectors.js]: Matching by assigned name could have some false positives, but the alternative would be pretty cumbersome I think (tracking assigned `connect` named import to its first function argument when invoked). Wondering if people would use [React's `this.setState`](https://facebook.github.io/react/docs/react-component.html#setstate) function callback form with `state` as the argument name.

File: docs/rules/redux-no-bound-selectors.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  getSite: getSite.bind( null, state ),
+} ) )( MyComponent );
+
+// or
+
+const mapState = ( state ) => ( {
+  getSite: ( id ) => getSite( state, id ),
+} );
+connect( mapState )( MyComponent );
+```
```
Comment (aduth) [inline docs/rules/redux-no-bound-selectors.md:22]: Are there any good "not warnings" examples we could include (as per convention with these docs) that might illustrate good alternatives?

**Review by aduth — COMMENTED** [review]
Lodash's `partial` could be similarly abused, though dunno that we care to address it since there's always going to be workarounds that we can't anticipate.

File: lib/rules/redux-no-bound-selectors.js (hunk 2 above)
Comment (mcsf) [reply lib/rules/redux-no-bound-selectors.js]: Good point with `setState`. I've just pushed dc0aecf to address that.

**Comment by mcsf** [comment]
Back from the meetup. 😄 

Another thing this catches is legitimate uses where state is recombined in a selector. Would you agree if we only reported a node if it is not wrapped in a `createSelector` call? Example caught by this rule:

```js
export const getSiteCollisions = createSelector(
	( state ) => {
		return map( filter( state.sites.items, ( wpcomSite ) => {
			const wpcomSiteUrlSansProtocol = withoutHttp( wpcomSite.URL );
			return ! wpcomSite.jetpack && some( state.sites.items, ( jetpackSite ) => {
				return (
					jetpackSite.jetpack &&
					wpcomSiteUrlSansProtocol === withoutHttp( jetpackSite.URL )
				);
			} );
		} ), 'ID' );
	},
	( state ) => state.sites.items
);
```

That way we also make sure that, aside from dodging the impact of repeated function binding via `connect`, we also recycle object references thanks to caching.

**Comment by aduth** [comment]
Rather than working to exempt cases, can we improve the intended targeting by testing only when bind occurs in context of `connect` ?

**Comment by mcsf** [comment]
> Rather than working to exempt cases, can we improve the intended targeting by testing only when bind occurs in context of connect ?

How would you deal with the common pattern of defining `mapState` as a
standalone/module function and passing it to `connect`? A simple test for a
couple of identifiers (e.g. `mapState`, `mapStateToProps`)? Or would you
attempt something more sophisticated, like tracking the variable passed to
`connect` and analyzing the corresponding function?

**Comment by aduth** [comment]
>Or would you attempt something more sophisticated, like tracking the variable passed to `connect` and analyzing the corresponding function?

This would seem to be best so far as limiting the number of false positives/negatives, but also the most difficult to implement.

**Comment by gwwar** [comment]
Just thinking of ideas, but it looks like we have access to the full ast tree, so we might break it down by:
- Verifying that we have a connect import from `react-redux`. And keeping a note of it's local name. Though it'll probably just  be `connect` in most cases.
- Find the `connect` callExpression
- Get the first argument
- If it's a function, run the checks
- if it's an identifier search for it in the tree and run the checks

**Comment by mcsf** [comment]
Alright, you two — I went with the longer route in 91596de. Added a test file showcasing examples. Let me know what you think! @gwwar @aduth

File: lib/rules/redux-no-bound-selectors.js [hunk 4]
```diff
@@ -15,59 +15,152 @@ const BIND_ERROR_MESSAGE = "Don't bind functions within `connect`." + REF;
...
+			'Program:exit': function() {
+				reports
+					.filter( ( { outer } ) =>
+						contains( nodesToReportOn, outer ) )
+					.forEach( ( { inner, message } ) =>
+						context.report( inner, message ) );
+			},
+
+			CallExpression: function( node ) {
+				if ( node.callee.name !== 'connect' ) {
```
Comment (gwwar) [inline lib/rules/redux-no-bound-selectors.js]: Should we also check in `ImportDeclaration` that we have a connect import from `react-redux`?

Comment (gwwar) [reply lib/rules/redux-no-bound-selectors.js]: Would it be worth walking `context.getSourceCode()` to see if we have a matching import? I think we'd be able to exit early in these handlers.

Comment (gwwar) [reply lib/rules/redux-no-bound-selectors.js]: Doing a quick search in Calypso, I think we can assume that it's probably from react-redux.

**Comment by gwwar** [comment]
Overall this appears to work really well. I'll let @aduth 👀  to see if any of this can be simplified. Besides that we'll want to 
- update the readme, to add some bad examples
- bump the version number
- update the main readme to point at the new rule
- update the changelog

File: lib/rules/redux-no-bound-selectors.js (hunk 4 above)
Comment (mcsf) [reply lib/rules/redux-no-bound-selectors.js]: > I think we can assume that it's probably from react-redux.

That was my thinking. If anyone ever does use something else called `connect`, they/we can deal with it accordingly.

File: lib/rules/redux-no-bound-selectors.js [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+				/*
+				 * The first argument of `connect` is expected to be either a
+				 * variable for a function, or an anonymous/inlined function.
+				 * Let's handle both cases.
+				 */
+				const mapStateNode = node.arguments[ 0 ];
+
+				let mapStateBody;
+
+				if ( mapStateNode.type === 'Identifier' ) {
```
Comment (aduth) [inline lib/rules/redux-no-bound-selectors.js]: Pushed a failing test caused by assumption of arguments at this line.

File: lib/rules/redux-no-bound-selectors.js [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+						definition.node.init.body;
+				} else {
+					/*
+					 * Inlined function? Easy peasy.
+					 */
+					mapStateBody = mapStateNode.body;
+				}
+
+				if ( mapStateBody ) {
+					nodesToReportOn.push( mapStateBody );
```
Comment (aduth) [inline lib/rules/redux-no-bound-selectors.js]: ~~Why do we wait 'til `Program:exit` to report on the node?~~

__Edit:__ Hmm, we need the whole set to know whether it's an issue? _Continues digesting_ 😄 
__Edit2:__ Yeah, disregard this one.

File: lib/rules/redux-no-bound-selectors.js [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+				}
+			},
+
+			Identifier: onIdentifierEnter,
```
Comment (aduth) [inline lib/rules/redux-no-bound-selectors.js]: I'm a little worried about the performance implications of an Identifier visitor. Do you think `CallExpression.callee.property.name` might be reasonable to check (and potentially more accurate) instead?

image

http://astexplorer.net/#/gist/1ce6f31efbd822108c1227fbb336098a/7c73e810541bec4f0f94dd486de21f49198ca697

**Comment by aduth** [comment]
Pushed another failing test relating to https://github.com/Automattic/eslint-plugin-wpcalypso/pull/38#discussion_r118779849 . Maybe a stretch, but testing any `bind` identifiers within `connect` seems a bit too overreaching.

**Comment by mcsf** [comment]
Thanks for the thorough input, @aduth. (I'll probably only work on this on Monday)

**Comment by mcsf** [comment]
Feedback addressed, with tests improved and satisfied. Think it's 🚢-able? @gwwar @aduth

**Comment by mcsf** [comment]
I've been catching (and hopefully preventing, but only out of luck by stumbling upon them) a couple of new occurrences of the anti-pattern in Calypso. Let's get this merged! :)

**Comment by gwwar** [comment]
I'm 👍  with the changes here. @aduth did you have any additional feedback?

**Review by aduth — APPROVED** [review]
To be clear, these two are valid issues? They're flagged for me in running the rule on Calypso's codebase. I'm guessing we're encouraging to create external selectors than mapping state within `connect`'s `mapStateToProps` ?

https://github.com/Automattic/wp-calypso/blob/f75c5ee/client/my-sites/theme/theme-features-card.js#L45
https://github.com/Automattic/wp-calypso/blob/f75c5ee/client/my-sites/themes/theme-options.js#L181-L185

Otherwise LGTM 👍

**Comment by mcsf** [comment]
Correct, @aduth, to me it would make sense to refactor those with memoized selectors (cc @ockham). Many thanks for the reviews, you two!

**Comment by mcsf** [comment]
(Aw, since this was my first merge in this repo the preset was to merge without squashing :/ sorry about that.)

**Pull request merged**